File: workers/test-basic/perl/workerTestBasic.pl
#!/bin/env perl
######################################################################
# Copyright (c) 2012, Yahoo! Inc. All rights reserved.
#
# This program is free software. You may copy or redistribute it under
# the same terms as Perl itself. Please see the LICENSE.Artistic file 
# included with this project for the terms of the Artistic License
# under which this project is licensed. 
######################################################################

use strict;
use warnings;

my $worker = Gearbox::WorkerTestBasicPerl->new( $ARGV[0] );
$worker->run();

package Gearbox::WorkerTestBasicPerl;

use strict;
use warnings;

use Gearbox::Worker;
use base 'Gearbox::Worker';
use JSON qw(from_json to_json);
use File::Basename qw(basename);

use constant DBDIR => "/usr/var/gearbox/db/test-basic-perl/";

sub new {
    my $class = shift;
    my $config = shift;

    my $self = $class->SUPER::new( $config );

    $self->register_handler( "do_get_testbasicperl_thing_v1" );
    $self->register_handler( "do_put_testbasicperl_thing_v1" ); 
    $self->register_handler( "do_post_testbasicperl_thing_v1" );
    $self->register_handler( "do_delete_testbasicperl_thing_v1" );

    return $self;
}

sub do_get_testbasicperl_thing_v1 {
    my $self = shift;
    my $job  = shift;
    my $resp = shift;

    my $environ = $job->environ();
    if ( exists $environ->{"TestBasic"} ) {
        $resp->add_header( "TestBasic", $environ->{"TestBasic"} );
    }

    my $args = $job->arguments();
    if( !@$args ) { # index GET
        my @files = glob(DBDIR . "*");
        my $out;

        # set things to an empty array in case our glob did not match anything
       $out->{things} = [];
       my $limit = 10; ;
       
       my $params = $job->query_params();
       if ( $params->{_count} ) {
           $limit = $params->{_count};
       }
       
       for( my $i=0; $i<@files && $i<$limit; $i++ ){ 
            my $matrix = $job->matrix_arguments();
            if ( $matrix->{_expand} && $matrix->{_expand} == 1 ) { 
                $out->{things}[$i] = from_json( slurp( $files[$i] ) );
            } else {
                $out->{things}[$i] = basename( $files[$i] );
            }
        }

        $resp->content( to_json( $out ) );

    } else {
        my $name = $args->[0];
        if ( -e DBDIR . $name ) {
            $resp->content( slurp( DBDIR . $name ) );
        } else {
            die new ERR_NOT_FOUND("thing \"$name\" not found");
        }
    }

    return $WORKER_SUCCESS;
}


sub do_put_testbasicperl_thing_v1 {
    my $self = shift;
    my $job = shift;
    my $resp = shift;

    # async message, so update status to let user know we are processing
    $resp->status()->progress(10);
    $resp->status()->add_message("processing");

    my $args = $job->arguments();
    if( !$args ) {
        die new ERR_BAD_REQUEST("missing required resource name");
    }

    my $in = from_json( $job->content() );
    if ( !$in->{id} ) { 
        die new ERR_BAD_REQUEST("missing required \"id\" field");
    }

    write_file( DBDIR . $args->[0], $job->content() );
 
    $resp->status()->add_message("done");

    return $WORKER_SUCCESS;
}

sub do_post_testbasicperl_thing_v1 {
    my $self = shift;
    my $job  = shift;
    my $resp = shift;

    $resp->status()->progress(10);
    $resp->status()->add_message("processing");
      
    if ( $job->operation() eq "create" ) {
        # post-create where the resource id is created for user
        # (instead of a PUT where the user specifies the name)

        # get the generated id
        my $in = from_json( $job->content() );
        $in->{id} = $job->resource_name();

        write_file( DBDIR . $job->resource_name(), to_json( $in ) );
    } else { 
        my $args = $job->arguments();
        # post update
        if( -e DBDIR . $args->[0] ) {
            my $in = from_json( $job->content() );
            my $out = from_json( slurp( DBDIR . $args->[0] ) );
            $out->{stuff} = $in->{stuff};
            write_file( DBDIR . $args->[0], to_json($out) );
        }
        else {
            die new ERR_NOT_FOUND("thing \"" . $args->[0] . "\" not found");
        }
    }

    $resp->status()->add_message("done");
    return $WORKER_SUCCESS;
}

sub do_delete_testbasicperl_thing_v1 {
    my $self = shift;
    my $job = shift;
    my $resp = shift;

    $resp->status()->progress(10);
    $resp->status()->add_message("processing");

    # don't actually delete if fake-out header is set
    my $headers = $job->headers();
    if( $headers->{"fake-out"} ) {
        $resp->status()->add_message("ignoring delete due to fake-out header");
        return $WORKER_SUCCESS;
    }
        
    my $args = $job->arguments();
    if( !$args ) {
        die new ERR_BAD_REQUEST("missing required resource name");
    }
      
    my $file = DBDIR . $args->[0];
    if ( -f $file ) {
      unlink($file);
    }
    else {
      die new ERR_NOT_FOUND("thing \"" . $args->[0] . "\" not found");
    }
        
    $resp->status()->add_message("done");
    return $WORKER_SUCCESS;
}

sub slurp {
    my $file = shift;
    return do {
        local $/ = undef;
        open my $fh, '<', $file or die "can't read contents of $file: $!\n";
        my $content = <$fh>;
        close $fh;
        $content;
    };
}

sub write_file {
    my $file = shift;
    my $contents = shift;
    open my $fh, ">", $file or die "Couldn't open $file to write: $!\n";

    print $fh $contents;
    close $fh;
    return;
}

1;

File: workers/test-basic/python/workerTestBasic.py
#!/usr/bin/env python

import glob
import json
import os
import sys

from gearbox import Worker, ERR_NOT_FOUND, ERR_BAD_REQUEST


def file_get_contents(filename):
    with open(filename, 'r') as f:
        contents = f.read()
    return contents


def file_put_contents(filename, contents):
    with open(filename, 'w') as f:
        f.write(contents)


class WorkerTestBasicPython(Worker):
    DBDIR = "/var/gearbox/db/test-basic-python/"

    def __init__(self, cfg):
        super(WorkerTestBasicPython, self).__init__(cfg)

        self.register_handler("do_get_testbasicpython_thing_v1")
        self.register_handler("do_put_testbasicpython_thing_v1")
        self.register_handler("do_post_testbasicpython_thing_v1")
        self.register_handler("do_delete_testbasicpython_thing_v1")

    def do_get_testbasicpython_thing_v1(self, job, resp):
        env = job.environ()
        if "TestBasic" in env:
            resp.add_header("TestBasic", env["TestBasic"])

        args = job.arguments()

        if not args:  # index GET
            files = glob.glob(self.DBDIR + "*")
            out = {"things": []}
            limit = len(files)

            if "_count" in job.query_params():
                cgi = job.query_params()
                limit = int(cgi["_count"])

            #TODO: this isn't using limit like perl/php does
            for i in xrange(limit):
                matrix = job.matrix_arguments()
                if "_expand" in matrix and matrix["_expand"] == "1":
                    contents = file_get_contents(files[i])
                    out["things"].append(json.loads(contents))
                else:
                    out["things"].append(os.path.basename(files[i]))

            resp.content(json.dumps(out))
        else:
            name = args[0]
            file_path = os.path.join(self.DBDIR, name)
            if os.path.exists(file_path):
                resp.content(file_get_contents(file_path))
            else:
                raise ERR_NOT_FOUND('thing "%(name)s" not found' % locals())

        return Worker.WORKER_SUCCESS

    def do_put_testbasicpython_thing_v1(self, job, resp):
        # async message, so update status to let user know we are processing
        resp.status().progress(10)
        resp.status().add_message("processing")
        args = job.arguments()
        if not args:
            raise ERR_BAD_REQUEST("missing required resource name")

        job_content = json.loads(job.content())
        if not "id" in job_content:
            raise ERR_BAD_REQUEST('missing required "id" field')

        filename = os.path.join(self.DBDIR, args[0])
        file_put_contents(filename, job.content())
        resp.status().add_message("done")

        return Worker.WORKER_SUCCESS

    def do_post_testbasicpython_thing_v1(self, job, resp):
        resp.status().progress(10)
        resp.status().add_message("processing")

        if job.operation() == "create":
            # post-create where the resource id is created for user
            # (instead of a PUT where the user specifies the name)

            # get the generated id
            job_content = json.loads(job.content())
            job_content["id"] = job.resource_name()

            filename = os.path.join(self.DBDIR, job.resource_name())
            file_put_contents(filename, json.dumps(job_content))
        else:
            args = job.arguments()
            # post update
            filename = os.path.join(self.DBDIR, args[0])
            if os.path.exists(filename):
                job_content = json.loads(job.content())
                out = json.loads(file_get_contents(filename))
                out["stuff"] = job_content["stuff"]
                file_put_contents(filename, json.dumps(out))
            else:
                raise ERR_NOT_FOUND('thing "%s" not found' % args[0])
        resp.status().add_message("done")
        return Worker.WORKER_SUCCESS

    def do_delete_testbasicpython_thing_v1(self, job, resp):
        resp.status().progress(10)
        resp.status().add_message("processing")

        # don't actually delete if fake-out header is set
        headers = job.headers()
        if "fake-out" in headers and int(headers["fake-out"]) == 1:
            resp.status().add_message("ignoring delete due to fake-out header")
            return Worker.WORKER_SUCCESS

        args = job.arguments()
        if not args:
            raise ERR_BAD_REQUEST("missing required resource name")

        filename = os.path.join(self.DBDIR, args[0])
        if os.path.exists(filename) and os.path.isfile(filename):
            os.unlink(filename)
        else:
            raise ERR_NOT_FOUND('thing "%s" not found' % args[0])

        resp.status().add_message("done")
        return Worker.WORKER_SUCCESS


if __name__ == "__main__":
    worker = WorkerTestBasicPython(sys.argv[2])
    worker.run()
